### ISSUE TITLE
annotations of ClassWithInitializer / ClassWithConstructors / ReferenceDeclaration

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-compiler#2160]
There are two parts to this.

First: make the following work:

```ceylon
final annotation class Foo() 
        satisfies OptionalAnnotation<Foo, ClassWithInitializerDeclaration> {}
annotation Foo foo() => Foo();
final annotation class Bar() 
        satisfies OptionalAnnotation<Bar, ClassWithConstructorsDeclaration> {}
annotation Bar bar() => Bar();

bar
class Z {
    new New() {}
}

foo
class W() {}
```

Second: make use of this in the Java annotation interop:

1. Any Java annotation marked `@Target(CONSTRUCTOR)` but not `@Target(CLASS)` gets the constraint `ConstructorDeclaration|ClassWithInitializerDeclaration`.
2. Any Java annotation marked `@Target(CLASS)` but not `@Target(CONSTRUCTOR)` gets the constraint `ClassDeclaration`.
2. Any Java annotation marked both `@Target({CLASS, CONSTRUCTOR})` gets the constraint `ClassWithConstructorsDeclaration|ConstructorDeclaration`.

Does that sound right to you @FroMage and @jvasileff?

See https://github.com/ceylon/ceylon-spec/issues/1294 and https://github.com/ceylon/ceylon.language/issues/680.

**Comment by jvasileff** [comment]
I think the breakdowns make sense.

For the third item, the synthetic `annotationType__CONSTRUCTOR` and `annotationType_TYPE` annotations will still be available for use on a ClassWithInitializerDeclaration, right?

**Comment by jvasileff** [comment]
I have to say, though, that I wish `class W() {}` and `class W { shared new () {} }` were made identical from a meta model perspective. The distinction seems to be a local concern.

**Comment by gavinking** [comment]
> For the third item, the synthetic `annotationType__CONSTRUCTOR` and `annotationType_TYPE` annotations will still be available for use on a ClassWithInitializerDeclaration, right?

Correct.  `annotationType__CONSTRUCTOR` applies to `ConstructorDeclaration|ClassWithInitializerDeclaration` and `annotationType_TYPE` applies to `ClassOrInterfaceDeclaration`.

> I have to say, though, that I wish `class W() {}` and `class W { shared new () {} }` were made identical from a meta model perspective. The distinction seems to be a local concern.

I disagree. From a metamodel perspective they are different, because in the second case I can have separate annotation lists for the `class` and the `new`.

**Comment by gavinking** [comment]
In addition, when https://github.com/ceylon/ceylon.language/issues/681 is done, we can resolve the ambiguities around Java annotations of value declarations:

1. Any Java annotation marked `@Target(FIELD)` but not `@Target(METHOD)` gets the constraint `ReferenceDeclaration`.
2. Any Java annotation marked `@Target(METHOD)` but not `@Target(FIELD)` gets the constraint `FunctionDeclaration`.
2. Any Java annotation marked both `@Target({METHOD, FIELD})` gets the constraint `FunctionDeclaration|ReferenceDeclaration`.

**Comment by FroMage** [comment]
I don't quite understand how that is better than having `ConstructorDeclaration` and `ClassDeclaration`? If every class with initialiser has an associated `ConstructorDeclaration` to represent its single initialiser/constructor callable, we always know where to place the annotation and where to retrieve it from.

**Comment by sgalles** [comment]
Just to be sure that I get this right : this does not work anymore because of this issue right ? is there a workaround in the meantime ?
```java
inject shared class ExampleServlet(Something constructorInjected) extends HttpServlet() {
```
`annotated program element does not satisfy annotation constraint: 'ClassWithInitializerDeclaration' is not assignable to 'FunctionOrValueDeclaration|ValueDeclaration|ConstructorDeclaration' ('FunctionOrValueDeclaration|ConstructorDeclaration')`

**Comment by jvasileff** [comment]
@sgalles I'm currently able to use `inject__CONSTRUCTOR` on actual constructors (but not on a class with init params).

**Comment by FroMage** [comment]
@gavinking my question above is still waiting for an answer.

**Comment by tombentley** [comment]
> Any Java annotation marked `@Target(METHOD)` but not `@Target(FIELD)` gets the constraint `FunctionDeclaration`.

I think that should be `FunctionDeclaration|ValueDeclaration` because someone in Javaland might write a constructor which they mean to be applied to getters only (for example). They have to annotate that annotation type with `@Target(METHOD)` because there's no `GETTER` (damn you Java for never making properties into a language level thing!), but then ceylon would prevent you from putting it on the value.

**Comment by tombentley** [comment]
Oh, and there is no `ElementType.CLASS` only `ElementType.TYPE`, so the mapping for classes and constructors should be, I think:

* If the annotation type contains both `ElementType.TYPE` and `ElementType.CONSTRUCTOR` then `ClassOrInterfaceDeclaration|AliasDeclaration|CallableConstructorDeclaration`
* If the annotation type contains `ElementType.TYPE` but not `ElementType.CONSTRUCTOR` then `ClassOrInterfaceDeclaration|AliasDeclaration`
* If the annotation type contains `ElementType.CONSTRUCTOR` but not `ElementType.TYPE` then `ClassWithInitializerDeclaration|CallableConstructorDeclaration`

**Comment by tombentley** [comment]
@sgalles @jvasileff stuff like

    inject shared class ExampleServlet(Something constructorInjected) extends HttpServlet() {

should now work without having to declare the class with constructors.

**Comment by tombentley** [comment]
Previously we've supported applying an Java annotation constrained to just `FIELD` to an `object` declaration (since it's unambiguous that the user wants the annotation on the underlying field) but since an object does not have a `ReferenceDeclaration` model this would no longer be possible.

**Comment by tombentley** [comment]
This constrained annotation interop stuff is a mess. The thing is that there are Ceylon constructs such as `object` declarations or value constructors which generate multiple annotatable things (such as a class, a constructor, a field and a getter method for `object`s, or a constructor a field and a getter method for value constructors). It's possible that someone has a java annotation which is constrained to several of these, say `{FIELD, METHOD, CONSTRUCTOR}` and wants to annotate the field. That's what we have the hackish `*__FIELD` disambiguation annotations for. The problem is when there's a constructor which is constrained to just `{FIELD}` it's still legitimate to want to annotate the field behind a value constructor or object, but because of the constraints we've applied that's no longer possible, because a `ValueConstructor` is not a `ReferenceDeclaration`.

In other words if we map the constrains in the "obvious" way then the annotation is too constrained to be used in special circumstances.

We could obviously map `{FIELD}` to `ReferenceDeclaration|ValueConstructorDeclaration`, but that doesn't help with the `object` case because we don't have an `ObjectDeclaration`.

**Comment by jvasileff** [comment]
> should now work without having to declare the class with constructors

Thanks @tombentley. I finally removed my workarounds and it works great.

**Comment by tombentley** [comment]
Hitherto when loading a Java `@interface` we've applied a mapping to its `@Target` and used that to construct an instantiation of `ConstrainedAnnotation` which we've then said is a superinterface of the `@interface` we're loading. The consequence of that is that all the annotation constructors (including the `__UGLY` ones) get the same constraints.

In reality there are several mappings of the `@Target`s. The "natural mapping" of something  `{FIELD}` is `ReferenceDeclaration`. But there's also a "maximal mapping", which includes the declaration metatype of every Ceylon element which can generate a field. I might want to apply a Java annotation to the static field generated by the transformation of a value constructor, for instance. To do that whatever the `foo__FIELD` annotation constructors type needs to inherit a different instantiation of `ConstrainedAnnotation`.

So... rather than having the @interface inherit from `ConstrainedAnnotation`, maybe we need to be generating `Foo__FIELD` annotation classes (in addition to the `foo__FIELD` annotation constructors) and have those classes inherit the `@interface` and also the relevant `ConstrainedAnnotation`, according to the maximal mapping. The `Foo` annotation class would be constrained according to the natural mapping.

That should keep the `__UGLY` annotations working, but mean that by default interop annotations get the most natural constraints.

**Comment by gavinking** [comment]
@tombentley Stop: don't do any work on this until I have time to think this through completely and specify what metamodel types the Java annotation constraints correspond to.

**Comment by gavinking** [comment]
So here's (finally) my proposal  for the mapping for Java annotation constraints:

| `ElementType`   | `Declaration` subtype                   |
|------------------|-----------------------------------------|
| `TYPE`          | `ClassOrInterfaceDeclaration`           |
| `FIELD`         | `ReferenceDeclaration`                  |
| `METHOD`        | `FunctionDeclaration|SetterDeclaration` |
| `CONSTRUCTOR`   | `CallableConstructorDeclaration` |
| `PACKAGE`       | `Package`                               |
| `MODULE`        | `Module`                                |

As a special case, if a Java annotation has `CONSTRUCTOR` and _not_ `TYPE`, then it gets `CallableConstructorDeclaration|ClassWithInitializerDeclaration`.

I *think* this proposal eliminates any ambiguities, while still letting you use all the useful Java annotations.

**Comment by gavinking** [comment]
The three (one major and two rather minor) holes in this are:

- There are `ReferenceDeclaration`s which are not `FIELD`s, for example parameters or local variables of functions. I *could* adjust the type checker so that these things get typed as `ValueDeclaration`s, I  supposed.
- Java has `PARAMETER` annotations, for example [`@Observes`](http://docs.oracle.com/javaee/6/api/javax/enterprise/event/Observes.html), which are unsupported. It's probably worth introducing some kind of `ParameterDeclaration` thingy to our metamodel. (Not for 1.2.0)
- If we introduced a `GetterDeclaration`, then we could change the mapping for `METHOD` to `FunctionDeclaration|GetterDeclaration|SetterDeclaration`, which might very occasionally be useful.

**Comment by gavinking** [comment]
Oh, oh, I got it, I know the fix for the "one major" problem:

> - There are `ReferenceDeclaration`s which are not `FIELD`s, for example parameters or local variables of functions. I *could* adjust the type checker so that these things get typed as `ValueDeclaration`s, I  supposed.

What we should do is add a second constraint on the _metamodel type_, constraining the annotation to `Attributes` only. I will update the table above to reflect this.

**Comment by tombentley** [comment]
@gavinking 

* I don't understand how model types come into this wrt constraints. That's just not something the typechecker supports. Please explain.
* I'd be surprised if there do not exist `METHOD`-constrained annotation types which their authors indend to annotate getters (sure they might not exist in Java EE, but they will exist). I find it odd that people won't be able to annotate their attributes with them. Sure I suppose `GetterDeclaration` is *a* solution, but the fact that Java annotations are driving the Ceylon metamodel seems wrong to me.
* Java has `LOCAL_VARIABLE` and `ANNOTATION_TYPE` too. (Also, there isn't a `MODULE`).
* You've not mentioned the `Foo__GETTER` annotations at all, so I don't know how they fit into your thinking.

**Comment by gavinking** [comment]
> That's just not something the typechecker supports.

Yes it does. It's a new feature in 1.2.

**Comment by gavinking** [comment]
> - I'd be surprised if there do not exist METHOD-constrained annotation types which their authors indend to annotate getters (sure they might not exist in Java EE, but they will exist).

Right, sure, but in a pinch, nothing's stopping you from writing this:

```ceylon
Type attribute => "blah";

javaAnnotation
Type getAttribute() => attribute;
```

It's good enough _for now_.

> - Java has `LOCAL_VARIABLE` and `ANNOTATION_TYPE` too. (Also, there isn't a `MODULE`).

I'm aware of that, but I'm pretty sure we don't need to support either of them right now.

> - You've not mentioned the `Foo__GETTER` annotations at all, so I don't know how they fit into your thinking.

I think they're disgustingly ugly, and, unless I'm missing something. that the need for them is almost completely alleviated by the above proposal. We can keep them for backward compatibility.

**Comment by gavinking** [comment]
P.S. @tombentley if you _insist_, I'm ok with adding a second special case:

- if a Java annotation has `METHOD` but _not_ `FIELD`, then it gets the type `FunctionDeclaration|SetterDeclaration|ValueDeclaration`.

That would be fine by me.

**Comment by tombentley** [comment]
> Yes it does. It's a new feature in 1.2.

Oh god, when did you sneak *that* in?

**Comment by tombentley** [comment]
> Right, sure, but in a pinch, nothing's stopping you from writing this:
> 
> ```ceylon
> Type attribute => "blah";
>
> javaAnnotation
> Type getAttribute() => attribute;
> ```

Nothing except the fact that the Ceylon `getAttribute()` gets quoted to the Java `$getAttribute()` which (stupidly in hindsight) is not a Java Bean getter.

**Comment by gavinking** [comment]
> Nothing except the fact that the Ceylon `getAttribute()` gets quoted to the Java `$getAttribute()` which (stupidly in hindsight) is not a Java Bean getter.

Doesn't matter. The Java annotation is expecting a `METHOD`. It doesn't say it has to be a getter.

**Comment by gavinking** [comment]
> Oh god, when did you sneak _that_ in?

_Sneak?_

The issue was opened in August 2013, and closed this May. You commented on it twice.

https://github.com/ceylon/ceylon-spec/issues/760

**Comment by tombentley** [comment]
Yes, but in this scenario the consumer of the Java annotation is wanting/needing it on a Java Bean getter.

**Comment by gavinking** [comment]
Maybe. I still think this is good enough for now.

**Comment by jvasileff** [comment]
To the extent I understand this proposal, I agree with @tombentley's points. Or, put another way, I don't see much changing.

In practice, I'll continue to need `inject__SETTER` for `javax.inject.Inject`, and I'll continue to want/need `max__GETTER` for `javax.validation.constraints.Max`, in order to preserve the property abstraction, and not try to program with fields, like an animal, when it's even possible.

I always thought the suffixes were useful, and just temporarily ugly until a better syntax is decided upon. And I continue to believe that there should be *at least* warnings when ambiguities exist - https://github.com/ceylon/ceylon-compiler/issues/2154.

**Comment by gavinking** [comment]
> In practice, I'll continue to need `inject__SETTER` for `javax.inject.Inject`

No. You definitely won't.

**Comment by jvasileff** [comment]
Setter will be inferred?

**Comment by gavinking** [comment]
If you put the annotation on a setter, then yes. OTOH, if you put it on a reference, then you will get direct field access. Which is perfectly fine, since you're just skipping a trivial noop setter.

**Comment by jvasileff** [comment]
Direct field access sucks. It breaks `late`.

**Comment by gavinking** [comment]
Well I'm not sure you need `late` here anymore. Ceylon classes have default constructors now, which Java frameworks like Hibernate and CDI can use.

**Comment by gavinking** [comment]
> in order to preserve the property abstraction, and not try to program with fields, like an animal, when it's even possible

Well, look, if you want to use CDI in a nonanimalistic fashion, you should be using constructor injection, don't you think?

**Comment by jvasileff** [comment]
Sure, constructor injection is good, but not always possible or convenient

**Comment by jvasileff** [comment]
> use CDI in a non animalistic fashion

Annotations are also used for data binding. I think you're designing this for a narrow definition of how things should work. Which I guess is ok, but doesn't always match what I want to do.

And, if they are useless, why even have attributes in Ceylon?

**Comment by gavinking** [comment]
Wait wait wait. This isn't going to be quite _perfect_ because we're doing Java interop here. Points of discomfort when using _Java_ annotations on Ceylon classes do not carry over to using _Ceylon_ annotations on Ceylon classes.

**Comment by jvasileff** [comment]
Well, I guess that's my point too. Acknowledge the imperfection by allowing (forcing) the programmer to translate/disambiguate, rather than guessing.

In the end, though, the only *critical* issue is to continue to allow and support `x__GETTER` and `x__SETTER`. I might not get the warnings/errors I would like, but at least I can still make things work.

**Comment by gavinking** [comment]
I'm not proposing to remove the `x__GETTER` stuff.

**Comment by gavinking** [comment]
P.S. One thing that I guess I probably should do is relax the constraint that says every class _must_ have either a constructor or an initializer. For use with `@Inject` I guess it's helpful to have a class that can't be constructed within the Ceylon language :-/

**Comment by jvasileff** [comment]
So no definite initialization check, and no need for `late`? Neat idea. Although, not good for testing.

**Comment by gavinking** [comment]
Well "good for testing" => constructor injection :-)

**Comment by tombentley** [comment]
I've persuaded the compiler to do something along the lines illustrated by @gavinking, but it means introducing an inconsistency: while a Ceylon annotation constrained to a `ReferenceDeclaration` (or `ValueDeclaration`) gets applied to the getter, a Java annotation gets applied to the setter (in order for `@Inject` to work). I guess this is an expected inconsistency, and for annotations which need to be applied to the getter we still have `foo__GETTER`.

**Comment by gavinking** [comment]
@tombentley after some reflection, I think I've changed my mind about this. I now think we should add Java-annotation-specific information to the model, and have the typechecker validate that. And perhaps just remove `ReferenceDeclaration`.

**Comment by gavinking** [comment]
I have implemented direct support for Java annotation constraints in the model and typechecker, without, of course, being able to test it. @tombentley needs to support this stuff in the Java model loader.

**Comment by tombentley** [comment]
I've made the compiler changes, still need to fix the runtime so we can find java annotations using `annotations()`

**Comment by FroMage** [comment]
Champagne!!!
